content/develop/concepts/architecture/threading.md

```diff
@@ -0,0 +1,167 @@
+        self.conn.send_ui_state(ui_state)
+        # on the other end of WebSocket connection,
+        # frontend receives the state and updates UI
+
+
+StreamlitServer().listen()
+```
+
+## `missing ScriptRunContext!` or `streamlit.errors.NoSessionContext`
+
+Since you are reading this page, chances are that you have already noticed such messages.
+
+Many Streamlit APIs, including `st.session_state` and multiple builtin widgets, expect themselves to run on a ScriptThread. Such APIs are typically related to per-session or per-page-run internal states.
+
+In a happy scenario, such code finds the `ScriptRunContext` object attached to the current thread (like in the illustriial code above). But when such Streamlit APIs couldn't, they issue such warnings or errors.
+
+## Custom threads
+
+An effective mitigation to delay, is to create threads and let them work concurrently. This works especially well with IO-heavy operations like remote query or data load.
+
+But due to the reasons you read by far, interacting with Streamlit code from your thread can be quirky. In this section we introduce 2 patterns to let different threads work together.
+
+Note: they are only patterns rather than complete solutions. You are advised to think them as an idea to start with. For example, one could extend pattern 1 into using a `concurrent.futures.ThreadPoolExecutor` thread pool.
+
+### 1. Only call Stramlit code from script thread
+
+Python threading provides ways to start a thread, wait for its execution, and collect its result. If we isolate custom thread from Streamlit APIs, everything should just work in order.
+
+In the following example page, `main` runs on the script thread and creates 2 custom `WorkerThread`. After WorkerThread-s run concurrently, `main` collects their results and updates UI.
+
+```py
+import streamlit as st
+import time
+from threading import Thread
+
+class WorkerThread(Thread):
+    def __init__(self, delay):
+        super().__init__()
+        self.delay = delay
+        self.return_value = None
+    def run(self):
+        # runs in custom thread, touches no Streamlit APIs
+        start_time = time.time()
+        time.sleep(self.delay)
+        end_time = time.time()
+        self.return_value = f"start: {start_time}, end: {end_time}"
+
+st.header("t1")
+result_1 = st.empty()
+st.header("t2")
+result_2 = st.empty()
+
+def main():
+    t1 = WorkerThread(5)
+    t2 = WorkerThread(5)
+    t1.start()
+    t2.start()
+    t1.join()
+    t2.join()
+    # main() runs in script thread, and can safely call Streamlit APIs
+    result_1.write(t1.return_value)
+    result_2.write(t2.return_value)
+
+main()
+
+```
+
+### 2. Expose context object to custom thread
+
+Alternatively, one can let a custom thread have access to the `ScriptRunContext` attached to ScriptThread. This pattern is also used by Streamlit standard widgets like [st.spinner](https://github.com/streamlit/streamlit/blob/develop/lib/streamlit/elements/spinner.py).
```

Comment:
Though I saw people are already going this way (in various GH issues), I'm not really sure about this pattern

1. it exposes internal object to page writers

2. it it less guaranteed. I don't know enough to say the probability where a `ScriptRunContext` suffices. The other pattern just looks "safer" to me because it assumes less.

Reply:
Officially, adding your own threads isn't supported. We don't want to include unsupported patterns in the main concepts area. This will likely go out as a Knowledge Base (KB) article. There is a subset of this information that can live in concepts, but it will need to be very carefully separated. For now, we can move this page to the KB so that we can get it published faster and I'll likely follow up with moving some of it back into concepts. (I still haven't read through any of it yet, so I still expect a few weeks before I get to this. Just a heads up about what will likely happen.)

In the longer run, the plan is to properly support multithreading and async tasks, but that work isn't currently on the schedule this quarter or next, so it wouldn't likely happen until next year at the earliest.

Reply:
That makes sense. I don't really want to promote the hack either.

When you come back, feel free to change or move things or ask me to do so 👍🏽